### ISSUE TITLE
noHeadings triggered even if role="heading" is set

### ISSUE THREAD

**Issue by robfentress** [ucfopen/UDOIT#99]
The noHeadings rule is triggered, even if role="heading" has been used to indicate that an element is to serve as a heading.  For example, this throws a warning:

```
<div role="heading" aria-level="1">I'm a level 1 heading</div>
<div role="heading" aria-level="2">I'm a level 2 heading</div>
<p>I suspect this will trigger a warning about adding headings, but we'll see.</p>
<p><img src="x.jpg" alt="The W3C" width="72" height="48" /></p>
<div role="heading" aria-level="2">I'm another level 2 heading</div>
<p>But I must explain to you how all this mistaken idea of denouncing pleasure and praising pain was born and I will give you a complete account of the system, and expound the actual teachings of the great explorer of the truth, the master-builder of human happiness. No one rejects, dislikes, or avoids pleasure itself, because it is pleasure, but because those who do not know how to pursue pleasure rationally encounter consequences that are extremely painful. Nor again is there anyone who loves or pursues or desires to obtain pain of itself, because it is pain, but because occasionally circumstances occur in which toil and pain can procure him some great pleasure. To take a trivial example, which of us ever undertakes laborious physical exercise, except to obtain some advantage from it? But who has any right to find fault with a man who chooses to enjoy a pleasure that has no annoying consequences, or one who avoids a pain that produces no resultant pleasure? On the other hand, we denounce with righteous indignation and dislike men who are so beguiled and demoralized by the charms of pleasure of the moment, so blinded by desire, that they cannot foresee.</p>
<p>But I must explain to you how all this mistaken idea of denouncing pleasure and praising pain was born and I will give you a complete account of the system, and expound the actual teachings of the great explorer of the truth, the master-builder of human happiness. No one rejects, dislikes, or avoids pleasure itself, because it is pleasure, but because those who do not know how to pursue pleasure rationally encounter consequences that are extremely painful. Nor again is there anyone who loves or pursues or desires to obtain pain of itself, because it is pain, but because occasionally circumstances occur in which toil and pain can procure him some great pleasure. To take a trivial example, which of us ever undertakes laborious physical exercise, except to obtain some advantage from it?</p>
<div role="heading" aria-level="2">I'm also a level 2 heading</div>
<p>But who has any right to find fault with a man who chooses to enjoy a pleasure that has no annoying consequences, or one who avoids a pain that produces no resultant pleasure? On the other hand, we denounce with righteous indignation and dislike men who are so beguiled and demoralized by the charms of pleasure of the moment, so blinded by desire, that they cannot foresee.</p>
<p><span>No one rejects, dislikes, or avoids pleasure itself, because it is pleasure, but because those who do not know how to pursue pleasure rationally encounter consequences that are extremely painful. Nor again is there anyone who loves or pursues or desires to obtain pain of itself, because it is pain, but because occasionally circumstances occur in which toil and pain can procure him some great pleasure.</span></p>
```

![screen shot 2016-02-24 at 4 01 22 pm](https://cloud.githubusercontent.com/assets/9356499/13300948/6eca625a-db10-11e5-975a-a5d3f3d01efd.png)

**Comment by AlanFCMV** [comment]
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/heading_role
> The best way to use this role is to not use it at all, and instead use the native heading tags \<h1> through \<h6>

**Comment by robfentress** [comment]
Indeed, it is true that ARIA should generally be avoided where possible, using role="heading" to indicate something is a heading is a technically valid solution and I don't think this should be considered an accessibility violation when used, especially given what appears to be be [universal support for this role](https://a11ysupport.io/tech/aria/heading_role) by screen readers.  I think to label it as a violation is misleading.  There may be edge cases where it is appropriate to use (else why even have it in ARIA?).

**Comment by bagofarms** [comment]
Thank you for the feedback @robfentress.  I do wonder why it is in ARIA, since it seems completely redundant.  While it is technically correct, it does not seem like a good practice, and I personally don't want to advocate for its use.  I guess this gets down to the purpose of UDOIT.  Should it advocate best practices and standards, or purely stick to standards?  Maybe we could add a suggestion that specifically looks for this attribute and suggests changing it to a heading tag?

**Comment by robfentress** [comment]
Yeah, I guess I'd just wan't there to be a warning rather than an error about stuff like this, since you don't know the circumstances behind a decision to use this role.  I can't think of a reason to make fake headings like this off the top of my head, and it almost certainly would never come up for content authors, as opposed to developers.  However, if people are using UDOIT to determine conformance to WCAG, for instance because they have a university policy committing them to that (as we do), it may give misleading results.

**Comment by robfentress** [comment]
Actually, looking at this a little more closely (it's been a while), I guess it isn't an "error", but a "suggestion".  Still, some sort of qualifier might be useful
